src/AnimationManager.ts

```diff
@@ -128,35 +113,183 @@ export class AnimationManager {
+		if (!retTrigger && this.axm.every(
+			userWish.destPos,
+			(v, opt) => isCircularable(v, opt.range, opt.circular as boolean[]))) {
+			console.warn("You can't stop the 'animation' event when 'circular' is true.");
+		}
+
+		if (retTrigger && !equal(userWish.destPos, depaPos)) {
+			const inputEvent = option && option.event || null;
+			this.animateLoop({
+				depaPos,
+				destPos: userWish.destPos,
+				duration: userWish.duration,
+				delta: this.axm.getDelta(depaPos, userWish.destPos),
+				isTrusted: !!inputEvent,
+				inputEvent,
+				input: option && option.input || null,
+			}, () => this.animationEnd());
+		}
+	}
+
+	public easing(p): number {
+		return p > 1 ? 1 : this._options.easing(p);
+	}
+
+	public setTo(pos: Axis, duration: number = 0) {
+		const axes: string[] = Object.keys(pos);
+		this.stop(axes);
+		const orgPos: Axis = this.axm.get(axes);
+
+		if (equal(pos, orgPos)) {
+			return this;
+		}
+		this.itm.setInterrupt(true);
+		let movedPos = filter(pos, (v, k) => orgPos[k] !== v);
+		if (!Object.keys(movedPos).length) {
+			return this;
+		}
+
+		movedPos = this.axm.map(movedPos, (v, opt) => {
+			const { range, circular } = opt;
+
+			if (circular && (circular[0] || circular[1])) {
+				return v;
+			} else {
+				return getInsidePosition(v, range, circular as boolean[]);
+			}
+		});
+
+		if (equal(movedPos, orgPos)) {
+			return this;
+		}
+
+		if (duration > 0) {
+			this.animateTo(movedPos, duration);
+		} else {
+			this.em.triggerChange(movedPos);
+			this.finish(false);
+		}
+
+		return this;
+	}
+
+	public setBy(pos: Axis, duration = 0) {
+		return this.setTo(
+			map(this.axm.get(Object.keys(pos)), (v, k) => v + pos[k]),
+			duration,
+		);
+	}
+
+	public updateAnimationPos(pos: Axis, restart: boolean): void {
+		const animateParam = this._animateParam;
+		if (!animateParam) {
+			return;
+		}
+		if (restart) {
+			this.setTo(pos, animateParam.duration - (new Date().getTime() - animateParam.startTime));
+		} else {
+			const currentPos = this.axm.get();
+			// When destination is changed, new delta should be calculated as remaining percent.
+			// For example, moving x:0, y:0 to x:200, y:200 and it has current easing percent of 92%. coordinate is x:184 and y:184
+			// If destination changes to x:300, y:300. xdelta:200, ydelta:200 changes to xdelta:116, ydelta:116 and use remaining easingPer as 100%, not 8% as previous.
+			// Therefore, original easingPer by time is kept. And divided by (1 - self._initialEasingPer) which means new total easing percent. Like calculating 8% as 100%.
+			this._initialEasingPer = this._prevEasingPer;
+			animateParam.delta = this.axm.getDelta(currentPos, pos);
+			animateParam.destPos = pos;
+		}
+	}
+
+	public updateAnimationDuration(duration: number,  restart: boolean): void {
+		const animateParam = this._animateParam;
+		if (!animateParam) {
+			return;
+		}
+		if (restart) {
+			this.setTo(animateParam.destPos, duration - (new Date().getTime() - animateParam.startTime));
```

Comment:
Is it tested for the case where the duration is less than the elapsed time?

Reply:
When duration is less than zero, animation ends Immediately at [setTo](https://github.com/malangfox/egjs-axes/blob/v3-add-animation-methods/src/AnimationManager.ts#L194-L199).

Reply:
If it is difficult to know whether the animation ends normally right away when a duration less than 0 is passed to `setTo`, I would like to fix this part more appropriately.

Reply:
There seems to be a problem with not doing anything. It should be one of the following actions.

1. Go directly to the location.
2. End the event without moving.